### PR TITLE
add 'Connect(ssh)' action for virtual machine

### PR DESCRIPTION
**Pull request by ivywei0125**
What does this implement/fix? Explain your changes.
--------------------------------------------------
add "connect(ssh)" action. The action is used to connect machine by ssh.
(action handler is not ready)

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java [hunk 1]
```diff
@@ -36,6 +37,11 @@ public void registerActions(AzureActionManager am) {
             .enabled(s -> s instanceof VirtualMachine && ((VirtualMachine) s).getFormalStatus().isRunning());
         am.registerAction(ADD_SSH_CONFIG, new Action<>(ADD_SSH_CONFIG, addSshConfigView));
 
+        final ActionView.Builder connectSshView = new ActionView.Builder("Connect(ssh)", AzureIcons.Action.RESTART.getIconPath())
```
Comment (Flanker32) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java]: restart as icons for connect?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.microsoft.azure.toolkit.intellij.vm.ssh;
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: please add copyright statement
```
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
```

Comment (ivywei0125) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: done

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java (hunk 1 above)
Comment (ivywei0125) [reply Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java]: already change to "terminal" icon

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    private static final String CMD_SSH = "ssh %s@%s";
+    private static final String CMD_SSH_KEY_PAIR = "ssh %s@%s -i %s";
+
+    public static void connectBySsh(VirtualMachine vm, @Nonnull Project project) {
+        final String machineName = vm.getName();
+        final SshConfig existingConfigs = SshConfigManager.getInstance(project)
+                .findConfigByName(String.format("Azure: %s", machineName));
+        final String userName = existingConfigs != null ? existingConfigs.getUsername() : vm.getAdminUserName();
+        final String hostIp = existingConfigs != null ?  existingConfigs.getHost() : vm.getHostIp();
+        final String privateKeyPath = defaultSshPrivateKeyPath();
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: -> `getDefaultSshPrivateKeyPath()` ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                } catch (IOException e) {
+                    AzureMessager.getMessager().error(e);
+                }
+            }));
+        }, AzureTask.Modality.ANY);
+    }
+
+    private static String defaultSshPrivateKeyPath() {
+        Path userHomePath = Paths.get(System.getProperty("user.home"));
+        Path filePath = Paths.get(userHomePath.toString(), ".ssh", "id_rsa");
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: Use `Paths.get(System.getProperty("user.home"),".ssh", "id_rsa" )` directly?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        AzureTaskManager.getInstance().runAndWait(() -> {
+            // create a new terminal tab
+            TerminalView terminalView = TerminalView.getInstance(project);
+            ShellTerminalWidget shellTerminalWidget = terminalView.createLocalShellWidget(null,
+                    String.format(SSH_TERMINAL_TABLE_NAME, machineName));
+            final AzureString messageTitle = description("webapp.open_ssh.app", machineName);
+            AzureTaskManager.getInstance().runInBackground(new AzureTask(project, messageTitle, false, () -> {
+                try {
+                    // create ssh connection in terminal
+                    if (existingConfigs != null) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: may be better to check whether there are private key configuration in existing configs and use config configuration if exists?

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java (hunk 1 above)
Comment (wangmingliang-ms) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-vm-lib/src/main/java/com/microsoft/azure/toolkit/ide/vm/VirtualMachineActionsContributor.java]: let's rename the action to `Connect Using SSH`.
refer to https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/factoryos/connect-using-ssh?view=windows-11.
please change the description accordingly

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 6]
```diff
@@ -1,13 +1,70 @@
...
+
+import static com.microsoft.azure.toolkit.lib.common.operation.OperationBundle.description;
+
+/**
+ * connect to selected machine by SSH in Intellij Ultimate.
+ */
 public class ConnectBySshAction {
+    private static final String SSH_TERMINAL_TABLE_NAME = "SSH - %s";
+    private static final String CMD_SSH = "ssh %s@%s";
+    private static final String CMD_SSH_KEY_PAIR = "ssh %s@%s -i %s";
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: it's OK to open the terminal manually in Community Edition.
we'd better reuse IntelliJ's own action (which is based on SSH configuration) in Ultimate Edition as we have discussed last Friday.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 7]
```diff
@@ -1,13 +1,70 @@
...
 
     public static void connectBySsh(VirtualMachine vm, @Nonnull Project project) {
-        // fixme empty handler @yuwe
+        final String machineName = vm.getName();
+        final SshConfig existingConfigs = SshConfigManager.getInstance(project)
+                .findConfigByName(String.format("Azure: %s", machineName));
+        final String userName = existingConfigs != null ? existingConfigs.getUsername() : vm.getAdminUserName();
+        final String hostIp = existingConfigs != null ?  existingConfigs.getHost() : vm.getHostIp();
+        final String privateKeyPath = defaultSshPrivateKeyPath();
+        AzureTaskManager.getInstance().runAndWait(() -> {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: why do we have to wait until it finished?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java [hunk 8]
```diff
@@ -1,13 +1,70 @@
...
+            // create a new terminal tab
+            TerminalView terminalView = TerminalView.getInstance(project);
+            ShellTerminalWidget shellTerminalWidget = terminalView.createLocalShellWidget(null,
+                    String.format(SSH_TERMINAL_TABLE_NAME, machineName));
+            final AzureString messageTitle = description("webapp.open_ssh.app", machineName);
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/ConnectBySshAction.java]: `webapp.*` is only for Azure Web App. please find or create another `vm.open_ssh.vm`.
this *key* (`webapp.open_ssh.app`) is also used in telemetry to track user actions.

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/icon/AzureIcons.java [hunk 9]
```diff
@@ -48,6 +48,7 @@ public static final class Action {
         public static final AzureIcon BROWSER = AzureIcon.builder().iconPath("/icons/action/browser").build();
         public static final AzureIcon PROPERTIES = AzureIcon.builder().iconPath("/icons/action/properties").build();
         public static final AzureIcon LOG = AzureIcon.builder().iconPath("/icons/action/log").build();
+        public static final AzureIcon TERMINAL = AzureIcon.builder().iconPath("/icons/Microsoft.Compute/actions/OpenTerminal_13x13.svg").build();
```
Comment (wangmingliang-ms) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/icon/AzureIcons.java]: there is no need to copy the icon to our repository, we can just use IntelliJ's icon directly in our code.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/PluginUtils.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package com.microsoft.azure.toolkit.intellij.common;
+
+public class PluginUtils {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/PluginUtils.java]: We may use `PlatformUtils` provided by IntelliJ directly
```
    PlatformUtils.isIdeaCommunity();
    PlatformUtils.isIdeaUltimate();
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/UltimateConnectBySshAction.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        SshTerminalCachingRunner runner = new SshTerminalCachingRunner(project, sshCredential, provider.getCharset());
+        ApplicationManager.getApplication().invokeLater((Runnable)(new Runnable() {
+            public final void run() {
+                connectToSshUnderProgress(project, runner, sshCredential);
+            }
+        }));
+    }
+
+    private static void connectToSshUnderProgress(final @NotNull Project project, final @NotNull SshTerminalCachingRunner runner, final @NotNull RemoteCredentials data) {
+        ProgressManager.getInstance().run(new Task.Backgroundable(project, RemoteSdkBundle.message("progress.title.connecting.to", new Object[]{runner.runningTargetName()}), true) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-vm/src/main/java/com/microsoft/azure/toolkit/intellij/vm/ssh/UltimateConnectBySshAction.java]: may migrate to `AzureTaskManager.getInstance().runInBackground()`

**Pull request merged**